frontend/src/concepts/pipelines/topology/parseUtils.ts

```diff
@@ -0,0 +1,151 @@
+  [componentName: string]: { [artifactId: string]: InputOutputArtifactType } | undefined;
+};
+export const parseComponentsForArtifactRelationship = (
+  components: PipelineComponentsKF,
+): ComponentArtifactMap =>
+  Object.entries(components).reduce<ComponentArtifactMap>(
+    (map, [componentId, componentValue]) =>
+      Object.entries(componentValue?.outputDefinitions?.artifacts ?? {}).reduce(
+        (artifactItems, [artifactId, value]) => {
+          const { artifactType } = value;
+
+          return {
+            ...artifactItems,
+            [componentId]: {
+              ...artifactItems[componentId],
+              [artifactId]: artifactType,
+            },
+          };
+        },
+        map,
+      ),
+    {},
+  );
+
+export type TaskArtifactMap = {
+  [taskName: string]: { outputArtifactKey: string; artifactId: string }[] | undefined;
+};
+export const parseTasksForArtifactRelationship = (tasks: DAG['tasks']): TaskArtifactMap =>
+  Object.values(tasks).reduce<TaskArtifactMap>(
+    (map, taskValue) =>
+      Object.entries(taskValue.inputs?.artifacts ?? {}).reduce(
+        (artifactItems, [artifactId, value]) => {
+          const { producerTask: taskId, outputArtifactKey } = value.taskOutputArtifact || {};
+          if (!taskId || !outputArtifactKey) {
+            // eslint-disable-next-line no-console
+            console.warn('Issue constructing artifact node', value);
+            return artifactItems;
+          }
+
+          return {
+            ...artifactItems,
+            [taskId]: [
+              ...(artifactItems[taskId] ?? []),
+              {
+                outputArtifactKey,
+                artifactId,
+              },
+            ],
+          };
+        },
+        map,
+      ),
+    {},
+  );
+
+export const parseInputOutput = (
+  definition?: InputOutputDefinition,
+): PipelineTaskInputOutput | undefined => {
+  let data: PipelineTaskInputOutput | undefined;
+  if (definition) {
+    const { artifacts, parameters } = definition;
+    data = {};
+
+    if (parameters) {
+      data = {
+        ...data,
+        params: Object.entries(parameters).map(([paramLabel, { parameterType }]) => ({
+          label: paramLabel,
+          type: parameterType,
+          // TODO: support value
+        })),
+      };
+    }
+
+    if (artifacts) {
+      data = {
+        ...data,
+        artifacts: Object.entries(artifacts).map(([paramLabel, { artifactType }]) => ({
+          label: paramLabel,
+          type: composeArtifactType(artifactType),
+          // TODO: support value
+        })),
+      };
+    }
+  }
+
+  return data;
+};
+
+export const parseRuntimeInfo = (
+  taskId: string,
+  runDetails?: RunDetailsKF,
+): PipelineTaskRunStatus | undefined => {
+  if (!runDetails) {
+    return undefined;
+  }
+
+  const { task_details: taskDetails } = runDetails;
+
+  const thisTaskDetail = taskDetails.find(({ display_name: name }) => name === `${taskId}-driver`);
```

Comment:
you only look at driver. I cannot get an error node to apear then. For some reason if a node has an error, it is not setting the state for the driver task, just the children ones

Reply:
Damn, seems I might need to find a bunch of them -- I didn't at the start, but it doesn't show active states if I don't.